cryostatio/cryostat3
# fix(connections): use Unis to enable internal connection retry behaviour

# Welcome to Cryostat3! 👋
## Before contributing, make sure you have:

**To recreate commits with GPG signature** `git fetch upstream && git rebase --force --gpg-sign upstream/main`
_______________________________________________

Related to #243
Related to #147

## Description of the change:
Refactors internal `TargetConnectionManager` implementation to use [`Uni`](https://smallrye.io/smallrye-mutiny/1.6.0/reference/uni-and-multi/). The unused `executeConnectedTaskAsync` is replaced with `executeConnectedTaskUni`, which executes on the current thread but in the reactive style. `executeConnectedTask`, which is used in many places and is generally how Cryostat performs remote operations on JMX or Agent HTTP targets, is refactored to call `executeConnectedTaskUni` internally. `executeDirect` is also refactored to behave similarly to `executeConnectedTaskUni`, but with the distinction that operations called through this method do not use the target connection cache, target connection locks, etc., so they are independent of any other operations or of discovery events. Both `executeDirect` and `executeConnectedTaskUni` are configured such that certain classes of exceptions trigger retries when caught, where other exceptions are allowed to propagate normally.

## Motivation for the change:
The "retry-on-certain-exceptions" behaviour allows Cryostat to internally re-attempt remote operations on targets when the operation fails due to certain exceptions which are known to happen sporadically and transiently. This increases Cryostat's reliability from the client's point of view, since rather than the client needing to send a new request to re-attempt after these failures, Cryostat will perform re-attempts on its own until either success or timeout.

## How to manually test:
1. `./smoketest.bash -Ogtr`
2. Open web UI and try various operations - starting recordings, creating snapshots, archiving recordings, deleting recordings, generating reports, defining automated rules, etc. Everything should continue working as normal. Sporadic (and hard to reproduce consistently) JMX connection failures should be reduced or eliminated. For example, try selecting Cryostat itself as the target, and try archiving the `onstart` recording that is already there. I occasionally see sporadic JMX connection failures when trying this before this PR.
